fuse_models/test/test_unicycle_2d_state_cost_function.cpp

```diff
@@ -0,0 +1,141 @@
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <fuse_models/unicycle_2d_state_cost_function.h>
+#include <fuse_models/unicycle_2d_state_cost_functor.h>
+
+#include <gtest/gtest.h>
+#include <fuse_core/eigen_gtest.h>
+
+#include <ceres/autodiff_cost_function.h>
+#include <ceres/gradient_checker.h>
+#include <Eigen/Dense>
+
+#include <limits>
+#include <vector>
+
+
+TEST(CostFunction, evaluateCostFunction)
+{
+  // Create cost function
+  const double process_noise_diagonal[] = { 1e-3, 1e-3, 1e-2, 1e-6, 1e-6, 1e-4, 1e-9, 1e-9 };
+  const fuse_core::Matrix8d covariance = fuse_core::Vector8d(process_noise_diagonal).asDiagonal();
+
+  const double dt{ 0.1 };
+  const fuse_core::Matrix8d sqrt_information{ covariance.inverse().llt().matrixU() };
+
+  const fuse_models::Unicycle2DStateCostFunction cost_function{ dt, sqrt_information };
+
+  // Evaluate cost function
+  const double position1[] = {0.0, 0.0};
+  const double yaw1[] = {0.0};
+  const double vel_linear1[] = {1.0, 0.0};
+  const double vel_yaw1[] = {1.570796327};
+  const double acc_linear1[] = {1.0, 0.0};
+
+  const double position2[] = {0.105, 0.0};
+  const double yaw2[] = {0.1570796327};
+  const double vel_linear2[] = {1.1, 0.0};
+  const double vel_yaw2[] = {1.570796327};
+  const double acc_linear2[] = {1.0, 0.0};
+
+  const double* parameters[] =
+  {
+    position1, yaw1, vel_linear1, vel_yaw1, acc_linear1,
+    position2, yaw2, vel_linear2, vel_yaw2, acc_linear2
+  };
+
+  fuse_core::Vector8d residuals;
+
+  const auto& block_sizes = cost_function.parameter_block_sizes();
+  const auto num_parameter_blocks = block_sizes.size();
+
+  const auto num_residuals = cost_function.num_residuals();
+
+  std::vector<fuse_core::MatrixXd> J(num_parameter_blocks);
+  std::vector<double*> jacobians(num_parameter_blocks);
+
+  for (size_t i = 0; i < num_parameter_blocks; ++i)
+  {
+    J[i].resize(num_residuals, block_sizes[i]);
+    jacobians[i] = J[i].data();
+  }
+
+  EXPECT_TRUE(cost_function.Evaluate(parameters, residuals.data(), jacobians.data()));
+
+  // We cannot use std::numeric_limits<double>::epsilon() tolerance because with the expected state2 above the residuals
+  // are not zero for position2.x = -4.389e-16 and yaw2 = -2.776e-16
+  EXPECT_MATRIX_NEAR(fuse_core::Vector8d::Zero(), residuals, 1e-15);
+
+  // Check jacobians are correct using a gradient checker
+  ceres::NumericDiffOptions numeric_diff_options;
+  ceres::GradientChecker gradient_checker(&cost_function, NULL, numeric_diff_options);
+
+  // We cannot use std::numeric_limits<double>::epsilon() tolerance because the worst relative error is 5.26356e-10
+  ceres::GradientChecker::ProbeResults probe_results;
+  // TODO(efernandez) probe_results segfaults when it's destroyed at the end of this TEST function, but Probe actually
```

Comment:
This is really unfortunate. It'd be great if you could give it a try in your system. I usually add a print a the end of the `TEST` to see if the segfault happens after, when objects are being destroyed.

My gut feeling is that this test should link against something else from `ceres`. I built and ran the `gradient_checker_test` from `ceres` and it didn't segfault, but then I took almost the same code from that test and tried it inside `fuse_models` and it crashed with the segfault described in the comment.

Reply:
I'll take a look at the gradient checker at some point, but I don't want to hold up this PR for it.